src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
-        uct_cuda_copy_iface_t *iface, uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
+static ucs_status_t
+uct_cuda_primary_ctx_push_first_active(CUdevice *cuda_device_p,
+                                       CUcontext *cuda_ctx_p)
+{
+    int num_devices, device_num;
+    ucs_status_t status;
+    CUdevice device;
+    CUcontext ctx;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGetCount(&num_devices));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    for (device_num = 0; device_num < num_devices; ++device_num) {
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGet(&device, device_num));
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        status = uct_cuda_primary_ctx_retain(device, 0, &ctx);
+        if (status == UCS_OK) {
+            break;
+        } else if (status == UCS_ERR_NO_DEVICE) {
+            continue;
+        } else {
+            return status;
+        }
+    }
+
+    if (device_num == num_devices) {
+        return UCS_ERR_NO_DEVICE;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxPushCurrent(ctx));
+    if (status != UCS_OK) {
+        UCT_CUDADRV_FUNC_LOG_WARN(cuDevicePrimaryCtxRelease(device));
+        return status;
+    }
+
+    *cuda_device_p = device;
+    *cuda_ctx_p    = ctx;
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_cuda_primary_ctx_pop_and_release(CUdevice cuda_device)
+{
+    if (ucs_likely(cuda_device == CU_DEVICE_INVALID)) {
+        return;
+    }
+
+    UCT_CUDADRV_FUNC_LOG_WARN(cuCtxPopCurrent(NULL));
+    UCT_CUDADRV_FUNC_LOG_WARN(cuDevicePrimaryCtxRelease(cuda_device));
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_copy_ctx_try_get_id(unsigned long long *ctx_id_p, CUdevice *device_p)
 {
     unsigned long long ctx_id;
     CUresult result;
-    khiter_t iter;
+    CUdevice device;
+    CUcontext ctx;
+    ucs_status_t status;
 
     result = uct_cuda_base_ctx_get_id(NULL, &ctx_id);
-    if (ucs_unlikely(result != CUDA_SUCCESS)) {
+    if (ucs_likely(result == CUDA_SUCCESS)) {
+        *ctx_id_p = ctx_id;
+        *device_p = CU_DEVICE_INVALID;
+        return UCS_OK;
+    }
+
+    status = uct_cuda_primary_ctx_push_first_active(&device, &ctx);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    result = uct_cuda_base_ctx_get_id(NULL, &ctx_id);
+    if (result != CUDA_SUCCESS) {
         UCT_CUDADRV_LOG(cuCtxGetId, UCS_LOG_LEVEL_ERROR, result);
+        uct_cuda_primary_ctx_pop_and_release(device);
         return UCS_ERR_IO_ERROR;
     }
 
+    *ctx_id_p = ctx_id;
+    *device_p = device;
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_copy_ctx_rsc_get(uct_cuda_copy_iface_t *iface, CUdevice *cuda_device_p,
+                          uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
+{
+    unsigned long long ctx_id;
+    CUdevice device;
+    ucs_status_t status;
+    khiter_t iter;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
+
+    status = uct_cuda_copy_ctx_try_get_id(&ctx_id, &device);
```

Comment:
maybe put the code of uct_cuda_copy_ctx_try_get_id here, since it seems to be used only here and both these functions are small now . so they can be united

Reply:
Combined the methods.